api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1728,202 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+	sclusterSboxMap := make(map[*Subcluster]*Sandbox)
+	for oldSubclusterName, oldSandboxName := range oldSubclusterInSandbox {
+		_, oldSubclusterInNewSandboxes := newSuclusterInSandbox[oldSubclusterName]
+		_, oldSubclusterInPersist := newSubclusterMap[oldSubclusterName]
+		// for unsandboxing, check shutdown field of the subcluster and sandbox
+		if !oldSubclusterInNewSandboxes && oldSubclusterInPersist {
+			oldSubcluster := oldSubclusterMap[oldSubclusterName]
+			oldSandbox := oldSandboxMap[oldSandboxName]
+			sclusterSboxMap[oldSubcluster] = oldSandbox
+		}
+	}
+	return sclusterSboxMap
+}
+
+// checkUnsandboxShutdownConditions ensures we will not unsandbox a subcluster that has shutdown set to true, or a subcluster
+// in a sandbox where the shutdown field for the sandbox is set to true
+func (v *VerticaDB) checkUnsandboxShutdownConditions(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	statusSClusterIndexMap := v.GenStatusSClusterIndexMap()
+	subclustersToUnsandbox := v.findSclustersToUnsandbox(old)
+	subclusterIndexMap := v.GenSubclusterIndexMap()
+	sandboxIndexMap := v.GenSandboxIndexMap()
+	for oldSubcluster, oldSandbox := range subclustersToUnsandbox {
+		oldSubclusterIndex, found := statusSClusterIndexMap[oldSubcluster.Name]
+		if oldSubcluster.Shutdown || found && v.Status.Subclusters[oldSubclusterIndex].Shutdown {
+			i := subclusterIndexMap[oldSubcluster.Name]
+			p := field.NewPath("spec").Child("subclusters").Index(i)
+			err := field.Invalid(p,
+				oldSubcluster.Shutdown,
+				fmt.Sprintf("cannot unsandbox subcluster %q that has Shutdown field set to true",
+					oldSubcluster.Name))
+			allErrs = append(allErrs, err)
+			continue
+		}
+		if oldSandbox.Shutdown {
+			i := sandboxIndexMap[oldSandbox.Name]
+			p := field.NewPath("spec").Child("sandboxes").Index(i)
+			err := field.Invalid(p,
+				oldSubcluster.Name,
+				fmt.Sprintf("cannot unsandbox subcluster %q in sandbox %q that has Shutdown field set to true",
+					oldSubcluster.Name, oldSandbox.Name))
+			allErrs = append(allErrs, err)
+			continue
+		}
+	}
+	return allErrs
+}
+
+// checkAnnotatedSubclustersInShutdownSandbox ensures: when a sandbox is to be shutdown, if a subcluster in it has annotation
+// vertica.com/shutdown-driven-by-sandbox set to true, the subcluster's shutdown field should not be updated
+func (v *VerticaDB) checkAnnotatedSubclustersInShutdownSandbox(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	newSandboxMap := v.GenSandboxMap()
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSubclusterIndexMap := v.GenSubclusterIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	oldSubclusterMap := oldObj.GenSubclusterMap()
+	for _, sandboxName := range persistSandboxes {
+		newSandbox := newSandboxMap[sandboxName]
+		oldSandbox := oldSandboxMap[sandboxName]
+		if newSandbox.Shutdown {
+			for _, subclusterName := range oldSandbox.Subclusters {
```

Comment:
Why are we checking old sandbox's subclusters? Doesn't this rule apply on the subclusters in new sandbox?

Reply:
This is from the JIRA: " If a sandbox has `shutdown`(spec) true and any of its subclusters has the annotation "vertica.com/shutdown-driven-by-sandbox" set to true in the old vdb, we should prevent the user from updating that subcluster's shutdown field. "
Here we check the subclusters that exist in both new and old vdbs.

Reply:
No we should check newSandbox. Here is the scenario this covers: sandbox sand1 has a subcluster sc1. you set sand1.shutdown to true, the operator will set sc1.shutdown to true and add the annotation to sc1. So now sand.shutdown = true, sc1.shutdown. true and sc1 has "vertica.com/shutdown-driven-by-sandbox". We do not want a user to come and change sc1.shutdown because that annotation is a proof that the sandbox drives the shutdown.

This is relevant only if sand1.shutdown is still true(newvdb)